### PR TITLE
Highlight underpacing insertion orders

### LINKED ISSUE #17: Update query page UI to allow user select the underpacing option [1 day]
Besides the input boxes, display a check box where users can select if they want to highlight the underpacing insertion orders in the spreadsheet.

Cost estimate: 1 day

### LINKED ISSUE #18: Update the service class to calculate whether IO is underpacing [1~3 day]
Cost estimate: 1~3 day

### LINKED ISSUE #19: Update the service class to highlight underpacing IOs in the spreadsheet [1~2 day]
Cost estimate: 1~2 day

### PR DESCRIPTION
**Pull request by hanj7221**
Closes #17, closes #18, closes #19.

Here is a screenshot illustrating how a sample underpacing insertion order will be highlighted:

Screen Shot 2020-07-14 at 9 08 12 AM

### CODE REVIEW

File: lib/src/excel.dart [hunk 1]
```diff
@@ -126,6 +137,32 @@ class ExcelDart {
     }));
   }
 
+  /// Highlight rows that are underpacing.
```
Comment (hanj7221) [inline lib/src/excel.dart:200]: Shall I change the relative notation used in this [function](https://github.com/googleinterns/dv360-excel-plugin/pull/53/commits/847e079cc3921677bce9ae2c909ac2f2e8fbb5d8#diff-5a9f11b722aad015b548cb227e5be190R118-R138) to match the regular notation instead? So instead of referencing RC[-2], we can use $M2 to reference the budget column instead. 

Both depends on hard coding though :<

File: lib/src/excel.dart [hunk 2]
```diff
@@ -19,6 +19,13 @@ class ExcelDart {
   static const _horizontalAlignment = 'Center';
   static const _borderStyle = 'Continuous';
   static const _currencyFormat = '\$#,##0.00';
+  static const _badFont = '#9C0006';
+  static const _badFill = '#FFC7CE';
+
+  static const _spentColumn = '\$N2';
```
Comment (hanj7221) [inline lib/src/excel.dart]: Add a comment to say that this has to match with the table header & body declaration.

File: lib/src/excel.dart [hunk 3]
```diff
@@ -19,6 +19,13 @@ class ExcelDart {
   static const _horizontalAlignment = 'Center';
   static const _borderStyle = 'Continuous';
   static const _currencyFormat = '\$#,##0.00';
+  static const _badFont = '#9C0006';
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: badFill/badFont kind of vague. consider something like underpaceFont

File: lib/src/excel.dart [hunk 4]
```diff
@@ -19,6 +19,13 @@ class ExcelDart {
...
+  static const _badFill = '#FFC7CE';
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: how were these values chosen? what color is this?

File: lib/src/excel.dart [hunk 5]
```diff
@@ -19,6 +19,13 @@ class ExcelDart {
...
+
+  static const _spentColumn = '\$N2';
+  static const _budgetColumn = '\$M2';
+  static const _startDateColumn = '\$O2';
+  static const _endDateColumn = '\$P2';
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: this will all change if you add something to the schema right? is there a way to calculate this value so changes to the schema will cascade without manual updates?

File: lib/src/excel.dart [hunk 6]
```diff
@@ -48,7 +55,8 @@ class ExcelDart {
   /// Waits for Office APIs to be ready and then creates
   /// a new spreadsheet with [sheetName] and populates the spreadsheet with
   /// entries in the [insertionOrderList].
-  void populate(List<InsertionOrder> insertionOrderList) async {
+  void populate(List<InsertionOrder> insertionOrderList,
+      bool highlightUnderpacing, String threshold) async {
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: when would highlighUnderpacing be false?

Comment (ruben-aghayan) [reply lib/src/excel.dart]: if highlight underpacing is false, does threshold do anything? do these make sense as optional arguments?

File: lib/src/excel.dart (hunk 1 above)
Comment (ruben-aghayan) [reply lib/src/excel.dart:200]: i think absolute > relative. But cant we calculate which column it is since we are specifying the schema? why hardcode?

File: lib/src/excel.dart [hunk 7]
```diff
@@ -126,6 +137,32 @@ class ExcelDart {
...
+  static void _highlightUnderpacingRows(
+      String tableName, String threshold) async {
+    await ExcelJS.run(allowInterop((context) async {
+      final table = context.workbook.tables.getItem(tableName);
+      final range = table.getDataBodyRange();
+      final format = range.conditionalFormats.add('Custom');
+
+      format.custom.rule.formula = '''
+      ($_spentColumn/$_budgetColumn) *
+      (DATEDIF($_startDateColumn, $_endDateColumn, "D") /
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: what kind of division is this? is it integer division? 
also what happens when the difference is 0?
(does the "D" mean double?) Since this is not a common library may be helpful to explain non-obvious args in a comment so they arent treated as "magic" inputs.

File: lib/src/query_component.dart [hunk 8]
```diff
@@ -15,20 +15,31 @@ import 'reporting_query_parser.dart';
            placeholder="Advertiser ID: 164337" debugId="advertiser-id-input">
     <input [(ngModel)]="insertionOrderId" 
            placeholder="Insertion Order ID: 8127549" debugId="io-id-input">
+    <br>
+    <input type="checkbox" [(ngModel)]="highlightUnderpacing"
+           debugId="underpacing" name="underpacing">
+    <label for="underpacing">Highlight underpacing insertion orders</label><br>
+    <input *ngIf="highlightUnderpacing" [(ngModel)]="underpacingThreshold"
```
Comment (ruben-aghayan) [inline lib/src/query_component.dart]: this is an interesting way to do it.
an alternative would be to have it as a cell in excel. (then a user can change it and it will automatically recalculate. does that work here?)

**Review by ruben-aghayan — CHANGES_REQUESTED** [review]
should any changes be reflected in the excel testing instructions?

File: lib/src/excel.dart (hunk 1 above)
Comment (hanj7221) [reply lib/src/excel.dart:200]: good point, good point, good point +1000.

File: lib/src/excel.dart (hunk 4 above)
Comment (hanj7221) [reply lib/src/excel.dart]: This mimics the 'bad' format in Excel.
Excel-bad-format

File: lib/src/excel.dart (hunk 6 above)
Comment (hanj7221) [reply lib/src/excel.dart]: optional arguments, yes

File: lib/src/excel.dart (hunk 4 above)
Comment (ruben-aghayan) [reply lib/src/excel.dart]: oh true. drop a comment to that effect!

File: lib/src/excel.dart (hunk 6 above)
Comment (hanj7221) [reply lib/src/excel.dart]: Actually, highlightUnderpacing will still remain as a required parameter, its value will be false if user doesn't check the checkbox. And threshold is no longer a parameter since it is changed to an excel cell input.

File: lib/src/excel.dart [hunk 9]
```diff
@@ -12,13 +12,36 @@ import 'util.dart';
...
-  static const _horizontalAlignment = 'Center';
+
+  static const _centerAlignment = 'Center';
+  static const _rightAlignment = 'Right';
+  static const _leftAlignment = 'Left';
+
   static const _borderStyle = 'Continuous';
+
   static const _currencyFormat = '\$#,##0.00';
+  static const _underpacingFormatFont = '#9C0006';
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: maybe comment that you are using the "bad" preset from excel?

File: lib/src/excel.dart [hunk 10]
```diff
@@ -48,22 +71,42 @@ class ExcelDart {
...
 
+        // Adds and formats underpacing info.
+        sheet.getRange('Query!A1:B2')
+          ..format.font.name = _fontName
+          ..format.font.size = _fontSize
+          ..format.font.bold = true
+          ..format.horizontalAlignment = _centerAlignment
+          ..values = [
+            ['highlight underpacing:', '$highlightUnderpacing'],
+            ['threshold: ', 0.9]
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: pull this out into a const

File: lib/src/excel.dart [hunk 11]
```diff
@@ -12,13 +12,36 @@ import 'util.dart';
 /// Service class that provides dart functions to interact with Excel.
 @Injectable()
 class ExcelDart {
-  static const _startAddress = 'A1';
-  static final _startAddressInt = 'A'.codeUnitAt(0);
+  static const _tableColumnStartAddress = 'A';
+  static const _tableRowStartAddress = 4;
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: how come it starts at row 4?

File: lib/src/excel.dart [hunk 12]
```diff
@@ -12,13 +12,36 @@ import 'util.dart';
...
+  static const _thresholdAddress = '\$B2';
```
Comment (ruben-aghayan) [inline lib/src/excel.dart:19]: can you include a screenshot? I think itll help illustrate the pr

File: lib/src/excel.dart [hunk 13]
```diff
@@ -12,13 +12,36 @@ import 'util.dart';
...
+  static const _underpacingFormatFill = '#FFC7CE';
+
+  /// Column addresses calculated based on [_tableHeader].
+  static final _spentColumn =
+      '\$${_getExcelColumnReference(_tableHeader.indexOf('Spent'))}'
+      '${_tableRowStartAddress + 1}';
+  static final _budgetColumn =
+      '\$${_getExcelColumnReference(_tableHeader.indexOf('Budget'))}'
+      '${_tableRowStartAddress + 1}';
+  static final _startDateColumn =
```
Comment (ruben-aghayan) [inline lib/src/excel.dart:53]: is this ok if the table doesnt exist yet (i.e. is it evaluated lazily?)

File: lib/src/excel.dart [hunk 14]
```diff
@@ -126,6 +175,44 @@ class ExcelDart {
     }));
   }
 
+  /// Highlight rows that are underpacing.
+  ///
+  /// Structured reference (i.e. Table[@Column]) is not allowed in Excel
+  /// conditional formatting. Regular reference (i.e. $A5) has to be used
+  /// instead. As table expands, Excel would automatically extends the
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: great comment
nit: will automatically extend

File: lib/src/excel.dart [hunk 15]
```diff
@@ -126,6 +175,44 @@ class ExcelDart {
...
+      // (spent / budget) * (flight duration/ current duration)
+      // if current duration or flight duration is zero days, then the formula
+      // simply calculates (spent / budget).
+      final formula = '''
+      IF(OR(DATEDIF($_startDateColumn, $_endDateColumn, "D") = 0, 
+            DATEDIF($_startDateColumn, NOW(), "D") = 0),
+         ($_spentColumn/$_budgetColumn),   
+         ($_spentColumn/$_budgetColumn) * 
+         (
+           DATEDIF($_startDateColumn, $_endDateColumn, "D") /
```
Comment (ruben-aghayan) [inline lib/src/excel.dart]: same question re:division. Is it integer division?

File: test/excel_test.dart [hunk 16]
```diff
@@ -8,8 +8,14 @@
...
+/// * Can check/uncheck the highlight underpacing option.
 /// * Click on populate
 ///
 /// Expected Result:
 /// * Should see a new sheet getting created that contains a table of insertion orders.
 /// * Insertion order data values should match with those on the DV360 UI.
+/// * Should see a cell displaying the underpacing threshold value.
+/// * If highlight underpacing is checked, underpacing insertion orders should be
+///   highlighted in red.
+/// * If highlight underpacing is checked, modifying the underpacing threshold value
```
Comment (ruben-aghayan) [inline test/excel_test.dart:23]: is there a way we can test that the correct rows are being highlighted using mocks?
e.g. verify that excel.format(underPace) font was called on rows 1,2,3 but not 4,5,6?

File: lib/src/excel.dart [hunk 17]
```diff
@@ -101,9 +144,15 @@ class ExcelDart {
...
+      if (highlightUnderpacing) _highlightUnderpacingRows(tableName);
     }));
   }
 
+  static String _getExcelColumnReference(int offset) =>
+      String.fromCharCode(_tableColumnStartAddress.codeUnitAt(0) + offset);
+
   /// Conditionally formats the Budget column based on Budget Unit.
   ///
   /// 'RC[-2]' in the formula follows Excel's Relative Notation and references
```
Comment (ruben-aghayan) [inline lib/src/excel.dart:178]: please create issue to refactor this from hardcoded relative to calculated absolute. (also elsewhere where applicable, and drop issue here.)

**Review by ruben-aghayan — CHANGES_REQUESTED** [review]

File: lib/src/excel.dart (hunk 12 above)
Comment (hanj7221) [reply lib/src/excel.dart:19]: Yep, sorry, that would make much more sense! See [here](https://github.com/googleinterns/dv360-excel-plugin/pull/53#issue-448467208)

File: lib/src/excel.dart (hunk 15 above)
Comment (hanj7221) [reply lib/src/excel.dart]: Ugh...sorry, forgot to include this in the comments. Nope, not integer division, just Excel regular division.

File: lib/src/excel.dart (hunk 17 above)
Comment (hanj7221) [reply lib/src/excel.dart:178]: sorry totally forgot to address this :(

File: lib/src/excel.dart (hunk 11 above)
Comment (hanj7221) [reply lib/src/excel.dart]: Please see [here](https://github.com/googleinterns/dv360-excel-plugin/pull/53#issue-448467208). Table starts on row 4, leaving some space to display underpacing options. 

If user doesn't check the underpacing checkbox, should we also display this, and shows 'highlight underpacing' as false, or should we hide this and move the table up?

File: lib/src/excel.dart (hunk 10 above)
Comment (hanj7221) [reply lib/src/excel.dart]: Can't turn this into a const since $highlightUnderpacing is a variable, but I will pull this out into a static final.

File: lib/src/excel.dart (hunk 13 above)
Comment (hanj7221) [reply lib/src/excel.dart:53]: Seems like so? it didn't complain or throw an error

File: test/excel_test.dart (hunk 16 above)
Comment (hanj7221) [reply test/excel_test.dart:23]: I'm afraid not, but we can talk about this :)

File: lib/src/excel.dart [hunk 18]
```diff
@@ -126,6 +197,46 @@ class ExcelDart {
...
+      // i.e. DATEDIF("7/1/2020", "7/31/2020", "D") = 30
+      //
+      // This formula calculates:
+      // (spent / budget) / (current duration / flight duration)
+      //
+      // Flight duration is calculated by (end date - start date + 1), and
+      // current duration is calculated by (now - start date + 1), so it is
+      // guaranteed that divide by zero will not happen.
+      final formula = ''' 
+       ($_spentColumn/$_budgetColumn) /
```
Comment (ruben-aghayan) [inline lib/src/excel.dart:225]: not sure about this formula. lets drop an issue here and immediately follow up with another pr and get arul's eyes on it. I think adding to the numerator and denominator will change the value. Lets get better granularity than days on the date calc, and not return answers when it has been in flight for <2 days or <10% of its flight (for e.g.) to guard against weird edge cases.

File: lib/src/excel.dart (hunk 14 above)
Comment (ruben-aghayan) [reply lib/src/excel.dart]: bump

File: test/excel_test.dart (hunk 16 above)
Comment (ruben-aghayan) [reply test/excel_test.dart:23]: lets drop an issue and immediately follow up. while we cant test the office api to excel ui connection, we may be able to test the excelDart to excel api connection.

**Review by ruben-aghayan — APPROVED** [review]

File: lib/src/excel.dart (hunk 11 above)
Comment (ruben-aghayan) [reply lib/src/excel.dart]: Looks good.

last thing, lets add the current pacing proportion into the schema
